pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,415 @@
+                             f'conduit_node_set_path_{value.dtype}_ptr')
+                fn(self, key, value.ctypes.data, value.size)
+            case _:
+                ValueError('ConduitNode: __setitem__ type not supported')
+
+    def __str__(self):
+        self.lib.conduit_node_print(self)
+        return '\n'
+
+    def append(self):
+        ptr = self.lib.conduit_node_append(self)
+        return ConduitNode(self.lib, ptr, child=True)
+
+
+class AscentWrappers(LibWrapper):
+    _libname = 'ascent_mpi'
+
+    # Functions
+    _functions = [
+        (None, 'ascent_close', c_void_p),
+        (c_void_p, 'ascent_create', c_void_p),
+        (None, 'ascent_execute', c_void_p, c_void_p),
+        (None, 'ascent_open', c_void_p, c_void_p),
+        (None, 'ascent_publish', c_void_p, c_void_p),
+    ]
+
+
+class AscentPlugin(RegionMixin, BasePlugin):
+    name = 'ascent'
+    systems = ['*']
+    formulations = ['dual', 'std']
+
+    # Element name mapping for conduit
+    cond_emap = {'pri': 'wedge', 'pyr': 'pyramid'}
+
+    # Ascent filters that implicitly require velocity expression
+    v_filter = ['gradient', 'qcriterion', 'vorticity']
+
+    def __init__(self, intg, cfgsect, suffix=None):
+        super().__init__(intg, cfgsect, suffix)
+
+        comm, rank, root = get_comm_rank_root()
+        self.nsteps = self.cfg.getint(cfgsect, 'nsteps')
+
+        # Get datatype
+        self.dtype = intg.system.backend.fpdtype
+
+        # Underlying system elements class
+        self.elementscls = intg.system.elementscls
+
+        # Load conduit library
+        self.conduit = ConduitWrappers()
+
+        # Expressions to plot and configs
+        self._exprs = defaultdict(dict)
+        self._fields_write = set()
+        self._fields_read = set()
+        self._init_fields(cfgsect)
+        self._init_scenes(cfgsect)
+        self._init_pipelines(cfgsect)
+
+        if not self._fields_read.issubset(self._fields_write):
+            raise AscentError('Not all fields used are defined')
+        
+        # Gradient pre-processing
+        self._init_gradients()
+
+        # Base directory for output and counter reset
+        basedir = self.cfg.getpath(cfgsect, 'basedir', '.', abs=True)
+
+        for scfg in self._scene_cfg.values():
+            scfg['path_gen'] = file_path_gen(basedir, scfg['image_name'],
+                                             intg.isrestart)
+            next(scfg['path_gen'])
+
+        # Generate conduit blueprint of mesh
+        self.mesh_n = ConduitNode(self.conduit)
+
+        self._ele_regions_lin = []
+        for ele in self._ele_regions:
+            if ele[1] == 'pyr':
+                divisor = 0
+            else:
+                order = self.cfg.getint('solver', 'order')
+                divisor = self.cfg.getint(cfgsect, 'division', order)
+
+            # Build the conduit blueprint mesh for the regions
+            self._build_blueprint(intg, rank, ele, divisor)
```

Comment:
Can avoid rank by using `intg.rallocs.prank` in the function itself, which may be more convenient as it uses the physical partition number rather than the MPI rank number.